src/gpu/intel/jit/ir/core.hpp

```diff
@@ -1752,10 +1852,51 @@ class var_t : public expr_impl_t {
     IR_DECLARE_TRAVERSERS()
 
     std::string name;
+    bool is_mutable = false;
 
 private:
-    var_t(const type_t &type, const std::string &name)
-        : expr_impl_t(_type_info(), type), name(name) {}
+    var_t(const type_t &type, const std::string &name, bool is_mutable)
+        : expr_impl_t(_type_info(), type), name(name), is_mutable(is_mutable) {}
+};
+
+class ref_t : public expr_impl_t {
+public:
+    IR_DECL_CORE_TYPE(ref_t)
+
+    static expr_t make(const expr_t &var, int off, int elems) {
+        return expr_t(new ref_t(var, off, elems));
+    }
+
+    bool is_equal(const object_impl_t &obj) const override {
+        if (!obj.is<self_type>()) return false;
+        auto &other = obj.as<self_type>();
+
+        return other.var.is_equal(var) && other.off == off
+                && other.elems == elems;
+    }
+
+    std::string str() const override {
+        std::ostringstream oss;
+        oss << var.str() << "[" << off;
+        if (elems > 1) oss << ":" << off + elems;
+        oss << "]";
+        return oss.str();
+    }
+
+    size_t get_hash() const override { return ir_utils::get_hash(str()); }
```

Comment:
This should be faster and aligned with the rest.
```suggestion
    size_t get_hash() const override { return ir_utils::get_hash(var, off, elems); }
```